### PR TITLE
Faster EpisodeStatisticsRecorder, converted from gym PR

### PR DESCRIPTION
**Pull request by DavidSlayback**
# Description

(Re-PR for gymnasium, this was reviewed and approved on the old gym repo [PR #3101](https://github.com/openai/gym/pull/3101)) 

This addresses two issues I've run into with RecordEpisodeStatistics

1) Silent list conversion in vectorized environments. Currently, RecordEpisodeStatistics converts terminated and truncated into lists, but obs and reward may still be arrays, and the new vectorized `info` interface uses arrays. I particularly run into this in my code if I do something like `dones.any()`. This PR leaves the underlying details of terminated/truncated alone by using NumPy array functionality.

2) Speed. Instead of looping through a potentially large batch of environments, this PR uses NumPy functions with their underlying vectorization to achieve the same task.

I considered building this with Jumpy, but I'm not sure on its status in the project, and it would also eat the cost of creating new arrays wherever I'm doing in-place modifications. 

I tested to make sure that this replicates the exact functionality of the original wrapper. To prove my speedup, I need to come up with some decent micro benchmarks; a couple attempts on Colab were fairly noisy.

## Type of change

- [x] Existing feature enhancement (non-breaking change which adds functionality)

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [N/A] I have commented my code, particularly in hard-to-understand areas
- [N/A] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [N/A] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by DavidSlayback** [comment]
Hmm...I'll take a look at the fail later today and fix it, it was a quick paste with a couple import changes.

**Comment by DavidSlayback** [comment]
Fixed, needed to store time as a vector (which seems a bit silly, it's the same time for all environments in a batch) and use float64 arrays for everything (to pass vector_list_info)

**Comment by DavidSlayback** [comment]
Weird, I never touched that file. Blacked it

File: gymnasium/wrappers/record_episode_statistics.py
Comment (gianlucadecola) [inline gymnasium/wrappers/record_episode_statistics.py:97]: Extremely minor, I think this could be the same as 
```python
np.full(self.num_envs, round(time.perf_counter() - self.t0, 6))
```
since self.num_envs is an integer it will be evaluated as the same shape without needing to wrap it in a tuple

**Review by gianlucadecola — APPROVED** [review]
Nice job! LGTM

File: gymnasium/wrappers/record_episode_statistics.py
Comment (RedTachyon) [inline gymnasium/wrappers/record_episode_statistics.py:99]: Why this change? Pretty much all the floating point math in gymnasium works on float32, this change will either add an extra cast somewhere along the line, or require an explicit cast by user by breaking their code (numpy doesn't care, but pytorch does)

**Comment by RedTachyon** [comment]
@pseudo-rnd-thoughts @gianlucadecola Just to be sure - did you go through the logic in both versions of the code to be sure that it's the same thing? It does seem cleaner which alone is sufficient reason for me to like this PR, but I'd need to sit down with it for a bit to make sure, and it might not be necessary. 

Also I see now that there's a message in the meantime about having to switch to float64 -- this is very weird and I'd be curious to see the reason for it, but I also hope that there's a different solution than using float64

File: gymnasium/wrappers/record_episode_statistics.py
Comment (DavidSlayback) [reply gymnasium/wrappers/record_episode_statistics.py:99]: Solely for passing the test for the VectorList wrapper, it checks each element with `assert isinstance(..., float)`, which only passed for float64 (not float32 or int). I agree I'd prefer to use float32, wasn't sure whether I could change pre-existing tests

Comment (RedTachyon) [reply gymnasium/wrappers/record_episode_statistics.py:99]: Can you point me to the specific test? File/line?

Comment (DavidSlayback) [reply gymnasium/wrappers/record_episode_statistics.py:99]: https://github.com/Farama-Foundation/Gymnasium/blob/9154d2db4e3955a99d81d1b1d982349cb78c80e6/tests/wrappers/test_vector_list_info.py#L56

Comment (RedTachyon) [reply gymnasium/wrappers/record_episode_statistics.py:99]: Is it the case that `list_info[i]["episode"][stats]` used to be a list of floats, and now is a numpy array? Or am I misunderstanding something here?

File: gymnasium/wrappers/record_episode_statistics.py
Comment (pseudo-rnd-thoughts) [inline gymnasium/wrappers/record_episode_statistics.py:96]: I'm not sure this will matter much but should be use `np.round` rather than the python `round` function given the value gets passed directly to a numpy array.

Secondly, this "t" value doesn't seem much help as `t0` is never updated to my eyes. So "t" seems to be the number of milli (or nano) seconds since the wrapper was initialised. 

This would break backward compatibility but would we want to change "t" to be the number of milli (or nano) seconds that the episode has taken. This would require us to have a `perf_counter` for every sub-environment but this shouldn't be too bad

File: gymnasium/wrappers/record_episode_statistics.py
Comment (pseudo-rnd-thoughts) [inline gymnasium/wrappers/record_episode_statistics.py:111]: Could you update `terminateds` to `termnations` and `trucnateds` to `truncations` as this is better english

File: gymnasium/wrappers/record_episode_statistics.py
Comment (pseudo-rnd-thoughts) [inline gymnasium/wrappers/record_episode_statistics.py:88]: For ease of reading, could you update to `dones = np.logical_or(termination, truncation)` as `|` is not clear what is means in this

File: gymnasium/wrappers/record_episode_statistics.py
Comment (pseudo-rnd-thoughts) [inline gymnasium/wrappers/record_episode_statistics.py:102]: Could we change this to
```python
if "episode" or "_episode" in infos:
    # raise warning
else:
    infos["episode"] = {...}
    if self.is_vector_env:
        infos["_episodes"] = ...
```
This seems better than the old way as it is a single element we are adding to the dictionary

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
@RedTachyon I believe that the implementation is the same. Could you check my suggestion on modifying "t" to be the episode length not time since initialisation. I suspect that very few people already use it so I think we can change it.
 
@DavidSlayback Looking at the code, given we are cleaning up the code, I have spotted a couple of other places that could be improved.

**Comment by DavidSlayback** [comment]
Made the requested changes.
1) I changed the storage arrays to float32 and int32. I had to adjust the test for vector_list_info to use np.isscalar. Basically, python isn't a big fan of float32. Getting a single value from a float64 array returns a python float, but from a float32 array, you get a numpy scalar.

2) `t0` is renamed to `episode_start_times` and tracks the start time for each episode. Pyright didn't like it being an `Optional[np.ndarray]`, hence the hack in the `__init__`. Updated wrapper documentation to reflect the change

3) Changed some of the wording for readability

**Review by pseudo-rnd-thoughts — APPROVED** [review]
There will be another Gym release with some minor bug fixes, hopefully, released in the next few days. 
To keep Gymnasium equivalent to Gym until launch Im going to wait for merge this

**Pull request merged**
